design/history/exploration-reports/2021.02-cross-language-fixtures.md

```diff
@@ -0,0 +1,160 @@
+let's talk about how we do fixtures
+===================================
+
+This is an exploration report to talk about ways we could do test fixtures that is cross-language and generally "feels nice".
+
+We have some work in this area already, but we need a lot more of it.
+Maybe some of it can adopt some standard conventions to save work and increase readability (and writeability).
+
+
+Requirements
+------------
+
+- separate document segments should be clearly delinated, and allow separate formats.
+	- (e.g., no, one homogeneus JSON document full of more JSON documents is not actually ideal.)
+- separate document segments should be able to be combined into one file and edited together.
+	- it just saves a lot of time and hastle in practice.
+	- avoids the question of "which of these files are related?" without spaming tons of directories (which would otherwise be common in practice).
+	- if document segments have to be split into wholy separate files, one can't copy-paste snippets of fixture easily.  (Means: bug reports and discussions in issues may be likely to invent an ad-hoc syntax for document segments, resulting in more work to turn those things into fixtures.)
+- easy to edit as a human.  Introducing new segments should ideally be context-free.
+	- writing these fixtures needs to be as low-friction as practically possible.
+	- e.g.: "put cursor at end of file, press enter, press dash three times" -- something like this complexity level is the goal.
+- should preserve whitespace precisely.
+	- it would be problematic for parser fixtures if whitespace isn't preserved!  (Mind touchy things like trailing linebreaks, especially.)
+- must be able to frame incomplete sub-documents, and not require knowledge of the parser for them.
+	- I need to use this to write fixtures for parsers _and their error modes_, which means sub-documents might be _intentionally_ unparsable.
+- nonrequirement: able to represent binary.
+	- as close as possible is certainly better.
+	- this can't be a goal simply because it runs directly crosswise to "easy to edit as a human".
+	- we may want to make a convention around base64 (?) and possible around comments in that format (?) ... but this can be implemented separately from the overall document container format.
```

Comment:
multibase might be nice here, could contain some flag that indicates the data is multibase encoded.

Reply:
I think the key on this point is that it's almost something we're going to next inside the container format.

E.g. precisely zero things we can pick up off the shelf are going to support multibase.

But to pick up something off the shelf, like txtar just to pick a random example, and then put either b64 or multibase inside it, is pretty trivial.